apps/emqx_durable_storage/src/emqx_ds_beamformer.erl

```diff
@@ -295,590 +418,952 @@ poll(Node, ReturnAddr, Shard, Iterator, Opts = #{timeout := Timeout}) ->
+disown_stuck_subscription(DBShard, SubTab, #sub_state{req_id = SubRef}) ->
+    [SubState] = ets:take(SubTab, SubRef),
+    gen_statem:cast(?via(DBShard), #disown_stuck_req{sub_state = SubState}).
+
+-spec disowned_sub_tab(dbshard()) -> ets:tid().
+disowned_sub_tab(DBShard) ->
+    #{stuck_sub_tab := TID} = gvar(DBShard),
+    TID.
+
+beams_conclude_node(DBShard, NextKey, BeamMaker, Node, BeamMakerNode) ->
+    #beam_builder{
+        cbm = CBM,
+        sub_tab = SubTab,
+        lagging = IsLagging,
+        queue_drop = QueueDrop,
+        queue_update = QueueUpdate
+    } = BeamMaker,
+    #beam_builder_node{
+        n_msgs = BatchSize,
+        pack = PackRev,
+        subs = Subs
+    } = BeamMakerNode,
+    Flags0 =
+        case IsLagging of
+            true -> ?DISPATCH_FLAG_LAGGING;
+            false -> 0
         end,
-    %% Execute callbacks:
-    OnMatch(MatchReqs),
-    %% Split matched requests by destination node:
-    ReqsByNode = maps:groups_from_list(
-        fun(#poll_req{node = Node}) -> Node end,
-        fun(#poll_req{return_addr = RAddr, it = It}) ->
-            {RAddr, It}
+    Destinations =
+        maps:fold(
+            fun(SubId, #beam_builder_sub{s = SubS0, mask = MaskEncoder, n_msgs = NMsgs}, Acc) ->
+                #sub_state{client = Client, it = It0} = SubS0,
+                ?tp(beamformer_fulfilled, #{sub_id => SubId}),
+                %% Update the iterator, start key and message
+                %% matcher for the subscription:
+                {ok, It} = update_iterator(CBM, DBShard, It0, NextKey),
+                %% Update the subscription state:
+                SubS = SubS0#sub_state{it = It, start_key = NextKey},
+                ets:insert(SubTab, SubS),
+                %% Update sequence number and check if the
+                %% subscription should remain active:
+                {Active, SeqNo} = keep_and_seqno(SubTab, SubS0, NMsgs),
+                %% Update the worker's queue:
+                case Active of
+                    true ->
+                        %% Subscription is keeping up with the
+                        %% acks. Update the worker's active
+                        %% queue:
+                        Flags = Flags0,
+                        QueueUpdate(
+                            SubS0,
+                            SubS
+                        );
+                    false ->
+                        %% This subscription is not keeping up
+                        %% with the acks. Freeze the
+                        %% subscription by dropping it from
+                        %% the worker's active queue:
+                        Flags = Flags0 bor ?DISPATCH_FLAG_STUCK,
+                        disown_stuck_subscription(DBShard, SubTab, SubS),
+                        QueueDrop(SubS0)
+                end,
+                DispatchMask = emqx_ds_dispatch_mask:enc_finalize(BatchSize + 1, MaskEncoder),
```

Comment:
Q: why the `+ 1` here, since (apparently) there's no message being appended to the batch?  Maybe add a comment explaining it?